lib/Controller/UsersController.php

```diff
@@ -469,6 +516,211 @@ public function create($username, $password, array $groups= [], $email='') {
 		);
 	}
 
+	/**
+	 * Set password for user using link
+	 *
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 *
+	 * @param string $token
+	 * @param string $userId
+	 * @return TemplateResponse
+	 */
+	public function setPasswordForm($token, $userId) {
+		try {
+			$this->checkPasswordSetToken($token, $userId);
+		} catch (UserTokenException $e) {
+			if ($e->getCode() === UserTokenException::TOKEN_TIME_EXPIRED) {
+				return new TemplateResponse(
+					'user_management', 'new_user/resendtokenbymail',
+					[
+						'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.resendToken', ['userId' => $userId])
+					], 'guest'
+				);
+			}
+			$this->log->logException($e, ['app' => 'user_management']);
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $e->getMessage()]]
+				], 'guest'
+			);
+		}
+
+		return new TemplateResponse(
+			'user_management', 'new_user/setpassword',
+			[
+				'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.setPassword', ['userId' => $userId, 'token' => $token])
+			], 'guest'
+		);
+	}
+
+	/**
+	 * @param string $token
+	 * @param string $userId
+	 * @return null
+	 * @throws InvalidUserTokenException
+	 * @throws UserTokenExpiredException
+	 * @throws UserTokenMismatchException
+	 */
+	private function checkPasswordSetToken($token, $userId) {
+		$user = $this->userManager->get($userId);
+
+		$splittedToken = \explode(':', $this->config->getUserValue($userId, 'owncloud', 'lostpassword', null));
+		if (\count($splittedToken) !== 2) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new InvalidUserTokenException($this->l10n->t('The token provided is invalid.'), UserTokenException::TOKEN_AND_TIME_INVALID);
+		}
+
+		if ($splittedToken[0] < ($this->timeFactory->getTime() - (int)$this->config->getAppValue('user_management', 'token_expire_time', '0')) ||
+			$user->getLastLogin() > $splittedToken[0]) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenExpiredException($this->l10n->t('The token provided had expired.'), UserTokenException::TOKEN_TIME_EXPIRED);
+		}
+
+		if (!\hash_equals($splittedToken[1], $token)) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenMismatchException($this->l10n->t('The token provided is invalid.'), UserTokenException::TOKEN_MISMACTH);
+		}
+	}
+
+	/**
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 * @param $userId
+	 */
+	public function resendToken($userId) {
+		$user = $this->userManager->get($userId);
+
+		if ($user === null) {
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $this->l10n->t('User: %s does not exist', [$userId])]]
```

Comment:
detailed error messages displayed on public pages are not good for security

this would allow an attacker to guess whether a given user exists or has an email set.

best is to have a generic error message that is the same regardless of the reason for failure (the reason could still be logged for the admin to see)

Reply:
if you check the current password reset code on stable10 you should see that we already had generic messages especially for that purpose